cmd/virt-launcher/virt-launcher.go

```diff
@@ -510,55 +510,37 @@ func ForkAndMonitor(containerDiskDir string) (int, error) {
 		return 1, err
 	}
 
-	exitStatus := make(chan syscall.WaitStatus, 10)
-	sigs := make(chan os.Signal, 10)
-	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGCHLD)
-	go func() {
-		for sig := range sigs {
-			switch sig {
-			case syscall.SIGCHLD:
-				var wstatus syscall.WaitStatus
-				wpid, err := syscall.Wait4(-1, &wstatus, syscall.WNOHANG, nil)
-				if err != nil {
-					log.Log.Reason(err).Errorf("Failed to reap process %d", wpid)
-				}
+	sigChan := make(chan os.Signal)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGCHLD)
 
-				// there's a race between cmd.Wait() and syscall.Wait4 when
-				// cleaning up the cmd's pid after it exits. This allows us
-				// to detect the correct exit code regardless of which wait
-				// wins the race.
-				if wpid == cmd.Process.Pid {
-					exitStatus <- wstatus
-				}
+	// wait for virt-launcher and collect the exit code
+	exitCode := 0
+	virtLauncherDone := false
+	for !virtLauncherDone {
+		switch sig := <-sigChan; sig {
+		case syscall.SIGCHLD:
+			var wstatus syscall.WaitStatus
+			wpid, err := syscall.Wait4(-1, &wstatus, syscall.WNOHANG, nil)
+			if err != nil {
+				log.Log.Reason(err).Errorf("Failed to reap process %d", wpid)
+			}
 
-			default:
-				log.Log.V(3).Log("signalling virt-launcher to shut down")
-				err := cmd.Process.Signal(syscall.SIGTERM)
-				sig.Signal()
-				if err != nil {
-					log.Log.Reason(err).Errorf("received signal %s but can't signal virt-launcher to shut down", sig.String())
-				}
+			if wpid == cmd.Process.Pid {
+				exitCode = wstatus.ExitStatus()
+				virtLauncherDone = true
 			}
-		}
-	}()
 
-	// wait for virt-launcher and collect the exit code
-	exitCode := 0
-	if err := cmd.Wait(); err != nil {
```

Comment:
What happens if the command exits but not because of one of the signals above (`syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGCHLD`)? Would there be a situation where the command crashes with no response due to us still waiting for `sigChan`?

Reply:
Hi! Thanks for the review!

The situation you describe it's not possible, regardless of how a process exits the parent always gets a SIGCHLD, if it doesn't then there's a kernel bug.

Reply:
Makes sense :) thanks